rest/blip_api_test.go

```diff
@@ -3628,3 +3628,113 @@ func TestMinRevPosWorkToAvoidUnnecessaryProveAttachment(t *testing.T) {
 	proveAttachmentAfter = btc.pushReplication.replicationStats.ProveAttachment.Value()
 	assert.Equal(t, proveAttachmentBefore, proveAttachmentAfter)
 }
+
+// Make sure that a client cannot open multiple subChanges subscriptions on a single blip context (SG #3222)
+//
+// - Open a one-off subChanges request, ensure it works.
+// - Open a subsequent continuous request, and ensure it works.
+// - Open another continuous subChanges, and asserts that it gets an error on the 2nd one, because the first is still running.
+// - Open another one-off subChanges request, assert we still get an error.
+func TestMultipleOutstandingChangesSubscriptions(t *testing.T) {
+	defer base.SetUpTestLogging(base.LevelInfo, base.KeyAll)()
+
+	bt, err := NewBlipTester(t)
+	require.NoError(t, err, "Error creating BlipTester")
+	defer bt.Close()
+
+	bt.blipContext.HandlerForProfile["changes"] = func(request *blip.Message) {
+		if !request.NoReply() {
+			// Send an empty response to avoid the Sync: Invalid response to 'changes' message
+			response := request.Response()
+			emptyResponseVal := []interface{}{}
+			emptyResponseValBytes, err := base.JSONMarshal(emptyResponseVal)
+			assert.NoError(t, err, "Error marshalling response")
+			response.SetBody(emptyResponseValBytes)
+		}
+	}
+
+	pullStats := bt.restTester.GetDatabase().DbStats.CBLReplicationPull()
+	assert.EqualValues(t, 0, pullStats.NumPullReplActiveContinuous.Value())
+	assert.EqualValues(t, 0, pullStats.NumPullReplActiveOneShot.Value())
+
+	// Open an initial continuous = false subChanges request, which we'd expect to release the lock after it's "caught up".
+	subChangesRequest := blip.NewRequest()
+	subChangesRequest.SetProfile("subChanges")
+	subChangesRequest.Properties["continuous"] = "false"
+	subChangesRequest.SetCompressed(false)
+	sent := bt.sender.Send(subChangesRequest)
+	goassert.True(t, sent)
+	subChangesResponse := subChangesRequest.Response()
+	goassert.Equals(t, subChangesResponse.SerialNumber(), subChangesRequest.SerialNumber())
+	errorCode := subChangesResponse.Properties["Error-Code"]
+	log.Printf("errorCode: %v", errorCode)
+	respBody, err := subChangesResponse.Body()
+	require.NoError(t, err)
+	assert.Equal(t, "", errorCode, "resp: %s", respBody)
+
+	pullStats = bt.restTester.GetDatabase().DbStats.CBLReplicationPull()
```

Comment:
We shouldn't need to keep retrieving this. It's a pointer, so can be set once at the start of the test, and values read from it throughout.